src/Processors/Executors/PipelineExecutor.cpp

```diff
@@ -265,6 +267,56 @@ void PipelineExecutor::initializeExecution(size_t num_threads)
 
     tasks.init(num_threads, profile_processors);
     tasks.fill(queue);
+
+    slots = ConcurrencyControl::instance().allocate(1, num_threads);
+    std::unique_lock lock{threads_mutex};
+    threads.reserve(num_threads);
+}
+
+void PipelineExecutor::spawnThreads()
+{
+    while (auto slot = slots->tryAcquire())
```

Comment:
If we don't have granted slots tryAcquire will return `{}`, i.e. `nullptr` containing `SlotPtr` and it will be interpreted as `false`.

Yeah, there is no condvar. I decided that periodic checking is okay, because:
 - it was hard to implement waiting on condvar: it requires some thread to be waiting
 - we anyway have at least one active thread that can do try to do upscaling by calling `spawnThreads()`
 - I expect overhead of periodic `tryAcquire()` calls to be low.